File: include/extras-attica/libdownloaddescription.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const downloaddescription_enums = enums;
const std = @import("std");

/// https://api.kde.org/attica-downloaddescription.html
pub const attica__downloaddescription = struct {
    /// New constructs a new Attica::DownloadDescription object.
    ///
    ///
    pub fn New() QtC.Attica__DownloadDescription {
        return qtc.Attica__DownloadDescription_new();
    }

    /// New2 constructs a new Attica::DownloadDescription object.
    ///
    /// ``` other: QtC.Attica__DownloadDescription ```
    pub fn New2(other: ?*anyopaque) QtC.Attica__DownloadDescription {
        return qtc.Attica__DownloadDescription_new2(@ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#operator-eq)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, other: QtC.Attica__DownloadDescription ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.Attica__DownloadDescription_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#id)
    ///
    /// ``` self: QtC.Attica__DownloadDescription ```
    pub fn Id(self: ?*anyopaque) i32 {
        return qtc.Attica__DownloadDescription_Id(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#type)
    ///
    /// ``` self: QtC.Attica__DownloadDescription ```
    ///
    /// Returns: ``` downloaddescription_enums.Type ```
    pub fn Type(self: ?*anyopaque) i32 {
        return qtc.Attica__DownloadDescription_Type(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#hasPrice)
    ///
    /// ``` self: QtC.Attica__DownloadDescription ```
    pub fn HasPrice(self: ?*anyopaque) bool {
        return qtc.Attica__DownloadDescription_HasPrice(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#category)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, allocator: std.mem.Allocator ```
    pub fn Category(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__DownloadDescription_Category(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::downloaddescription.Category: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#name)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, allocator: std.mem.Allocator ```
    pub fn Name(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__DownloadDescription_Name(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::downloaddescription.Name: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#link)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, allocator: std.mem.Allocator ```
    pub fn Link(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__DownloadDescription_Link(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::downloaddescription.Link: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#distributionType)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, allocator: std.mem.Allocator ```
    pub fn DistributionType(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__DownloadDescription_DistributionType(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::downloaddescription.DistributionType: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#priceReason)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, allocator: std.mem.Allocator ```
    pub fn PriceReason(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__DownloadDescription_PriceReason(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::downloaddescription.PriceReason: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#priceAmount)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, allocator: std.mem.Allocator ```
    pub fn PriceAmount(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__DownloadDescription_PriceAmount(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::downloaddescription.PriceAmount: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#size)
    ///
    /// ``` self: QtC.Attica__DownloadDescription ```
    pub fn Size(self: ?*anyopaque) u32 {
        return qtc.Attica__DownloadDescription_Size(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#gpgFingerprint)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, allocator: std.mem.Allocator ```
    pub fn GpgFingerprint(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__DownloadDescription_GpgFingerprint(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::downloaddescription.GpgFingerprint: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#gpgSignature)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, allocator: std.mem.Allocator ```
    pub fn GpgSignature(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__DownloadDescription_GpgSignature(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::downloaddescription.GpgSignature: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#packageName)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, allocator: std.mem.Allocator ```
    pub fn PackageName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__DownloadDescription_PackageName(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::downloaddescription.PackageName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#repository)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, allocator: std.mem.Allocator ```
    pub fn Repository(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__DownloadDescription_Repository(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::downloaddescription.Repository: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#tags)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, allocator: std.mem.Allocator ```
    pub fn Tags(self: ?*anyopaque, allocator: std.mem.Allocator) [][]const u8 {
        const _arr: qtc.libqt_list = qtc.Attica__DownloadDescription_Tags(@ptrCast(self));
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("attica::downloaddescription.Tags: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("attica::downloaddescription.Tags: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setId)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, id: i32 ```
    pub fn SetId(self: ?*anyopaque, id: i32) void {
        qtc.Attica__DownloadDescription_SetId(@ptrCast(self), @intCast(id));
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setType)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, typeVal: downloaddescription_enums.Type ```
    pub fn SetType(self: ?*anyopaque, typeVal: i32) void {
        qtc.Attica__DownloadDescription_SetType(@ptrCast(self), @intCast(typeVal));
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setHasPrice)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, hasPrice: bool ```
    pub fn SetHasPrice(self: ?*anyopaque, hasPrice: bool) void {
        qtc.Attica__DownloadDescription_SetHasPrice(@ptrCast(self), hasPrice);
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setCategory)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, category: []const u8 ```
    pub fn SetCategory(self: ?*anyopaque, category: []const u8) void {
        const category_str = qtc.libqt_string{
            .len = category.len,
            .data = category.ptr,
        };
        qtc.Attica__DownloadDescription_SetCategory(@ptrCast(self), category_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setName)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, name: []const u8 ```
    pub fn SetName(self: ?*anyopaque, name: []const u8) void {
        const name_str = qtc.libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        qtc.Attica__DownloadDescription_SetName(@ptrCast(self), name_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setLink)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, link: []const u8 ```
    pub fn SetLink(self: ?*anyopaque, link: []const u8) void {
        const link_str = qtc.libqt_string{
            .len = link.len,
            .data = link.ptr,
        };
        qtc.Attica__DownloadDescription_SetLink(@ptrCast(self), link_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setDistributionType)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, distributionType: []const u8 ```
    pub fn SetDistributionType(self: ?*anyopaque, distributionType: []const u8) void {
        const distributionType_str = qtc.libqt_string{
            .len = distributionType.len,
            .data = distributionType.ptr,
        };
        qtc.Attica__DownloadDescription_SetDistributionType(@ptrCast(self), distributionType_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setPriceReason)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, priceReason: []const u8 ```
    pub fn SetPriceReason(self: ?*anyopaque, priceReason: []const u8) void {
        const priceReason_str = qtc.libqt_string{
            .len = priceReason.len,
            .data = priceReason.ptr,
        };
        qtc.Attica__DownloadDescription_SetPriceReason(@ptrCast(self), priceReason_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setPriceAmount)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, priceAmount: []const u8 ```
    pub fn SetPriceAmount(self: ?*anyopaque, priceAmount: []const u8) void {
        const priceAmount_str = qtc.libqt_string{
            .len = priceAmount.len,
            .data = priceAmount.ptr,
        };
        qtc.Attica__DownloadDescription_SetPriceAmount(@ptrCast(self), priceAmount_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setSize)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, size: u32 ```
    pub fn SetSize(self: ?*anyopaque, size: u32) void {
        qtc.Attica__DownloadDescription_SetSize(@ptrCast(self), @intCast(size));
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setGpgFingerprint)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, fingerprint: []const u8 ```
    pub fn SetGpgFingerprint(self: ?*anyopaque, fingerprint: []const u8) void {
        const fingerprint_str = qtc.libqt_string{
            .len = fingerprint.len,
            .data = fingerprint.ptr,
        };
        qtc.Attica__DownloadDescription_SetGpgFingerprint(@ptrCast(self), fingerprint_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setGpgSignature)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, signature: []const u8 ```
    pub fn SetGpgSignature(self: ?*anyopaque, signature: []const u8) void {
        const signature_str = qtc.libqt_string{
            .len = signature.len,
            .data = signature.ptr,
        };
        qtc.Attica__DownloadDescription_SetGpgSignature(@ptrCast(self), signature_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setPackageName)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, packageName: []const u8 ```
    pub fn SetPackageName(self: ?*anyopaque, packageName: []const u8) void {
        const packageName_str = qtc.libqt_string{
            .len = packageName.len,
            .data = packageName.ptr,
        };
        qtc.Attica__DownloadDescription_SetPackageName(@ptrCast(self), packageName_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setRepository)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, repository: []const u8 ```
    pub fn SetRepository(self: ?*anyopaque, repository: []const u8) void {
        const repository_str = qtc.libqt_string{
            .len = repository.len,
            .data = repository.ptr,
        };
        qtc.Attica__DownloadDescription_SetRepository(@ptrCast(self), repository_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setTags)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, tags: [][]const u8, allocator: std.mem.Allocator ```
    pub fn SetTags(self: ?*anyopaque, tags: [][]const u8, allocator: std.mem.Allocator) void {
        var tags_arr = allocator.alloc(qtc.libqt_string, tags.len) catch @panic("attica::downloaddescription.SetTags: Memory allocation failed");
        defer allocator.free(tags_arr);
        for (tags, 0..tags.len) |item, i| {
            tags_arr[i] = .{
                .len = item.len,
                .data = item.ptr,
            };
        }
        const tags_list = qtc.libqt_list{
            .len = tags.len,
            .data = tags_arr.ptr,
        };
        qtc.Attica__DownloadDescription_SetTags(@ptrCast(self), tags_list);
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#version)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, allocator: std.mem.Allocator ```
    pub fn Version(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__DownloadDescription_Version(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::downloaddescription.Version: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-downloaddescription.html#setVersion)
    ///
    /// ``` self: QtC.Attica__DownloadDescription, version: []const u8 ```
    pub fn SetVersion(self: ?*anyopaque, version: []const u8) void {
        const version_str = qtc.libqt_string{
            .len = version.len,
            .data = version.ptr,
        };
        qtc.Attica__DownloadDescription_SetVersion(@ptrCast(self), version_str);
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.Attica__DownloadDescription ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.Attica__DownloadDescription_Delete(@ptrCast(self));
    }
};

/// https://api.kde.org/attica-downloaddescription.html#types
pub const enums = struct {
    pub const Type = enum {
        pub const FileDownload: i32 = 0;
        pub const LinkDownload: i32 = 1;
        pub const PackageDownload: i32 = 2;
    };
};

File: src_original/extras-attica/libdownloaddescription.cpp
#define WORKAROUND_INNER_CLASS_DEFINITION_Attica__DownloadDescription
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <downloaddescription.h>
#include "libdownloaddescription.h"
#include "libdownloaddescription.hxx"

Attica__DownloadDescription* Attica__DownloadDescription_new() {
    return new Attica::DownloadDescription();
}

Attica__DownloadDescription* Attica__DownloadDescription_new2(const Attica__DownloadDescription* other) {
    return new Attica::DownloadDescription(*other);
}

void Attica__DownloadDescription_OperatorAssign(Attica__DownloadDescription* self, const Attica__DownloadDescription* other) {
    self->operator=(*other);
}

int Attica__DownloadDescription_Id(const Attica__DownloadDescription* self) {
    return self->id();
}

int Attica__DownloadDescription_Type(const Attica__DownloadDescription* self) {
    return static_cast<int>(self->type());
}

bool Attica__DownloadDescription_HasPrice(const Attica__DownloadDescription* self) {
    return self->hasPrice();
}

libqt_string Attica__DownloadDescription_Category(const Attica__DownloadDescription* self) {
    QString _ret = self->category();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Attica__DownloadDescription_Name(const Attica__DownloadDescription* self) {
    QString _ret = self->name();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Attica__DownloadDescription_Link(const Attica__DownloadDescription* self) {
    QString _ret = self->link();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Attica__DownloadDescription_DistributionType(const Attica__DownloadDescription* self) {
    QString _ret = self->distributionType();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Attica__DownloadDescription_PriceReason(const Attica__DownloadDescription* self) {
    QString _ret = self->priceReason();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Attica__DownloadDescription_PriceAmount(const Attica__DownloadDescription* self) {
    QString _ret = self->priceAmount();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

unsigned int Attica__DownloadDescription_Size(const Attica__DownloadDescription* self) {
    return static_cast<unsigned int>(self->size());
}

libqt_string Attica__DownloadDescription_GpgFingerprint(const Attica__DownloadDescription* self) {
    QString _ret = self->gpgFingerprint();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Attica__DownloadDescription_GpgSignature(const Attica__DownloadDescription* self) {
    QString _ret = self->gpgSignature();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Attica__DownloadDescription_PackageName(const Attica__DownloadDescription* self) {
    QString _ret = self->packageName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Attica__DownloadDescription_Repository(const Attica__DownloadDescription* self) {
    QString _ret = self->repository();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_list /* of libqt_string */ Attica__DownloadDescription_Tags(const Attica__DownloadDescription* self) {
    QList<QString> _ret = self->tags();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QString _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void Attica__DownloadDescription_SetId(Attica__DownloadDescription* self, int id) {
    self->setId(static_cast<int>(id));
}

void Attica__DownloadDescription_SetType(Attica__DownloadDescription* self, int typeVal) {
    self->setType(static_cast<Attica::DownloadDescription::Type>(typeVal));
}

void Attica__DownloadDescription_SetHasPrice(Attica__DownloadDescription* self, bool hasPrice) {
    self->setHasPrice(hasPrice);
}

void Attica__DownloadDescription_SetCategory(Attica__DownloadDescription* self, const libqt_string category) {
    QString category_QString = QString::fromUtf8(category.data, category.len);
    self->setCategory(category_QString);
}

void Attica__DownloadDescription_SetName(Attica__DownloadDescription* self, const libqt_string name) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    self->setName(name_QString);
}

void Attica__DownloadDescription_SetLink(Attica__DownloadDescription* self, const libqt_string link) {
    QString link_QString = QString::fromUtf8(link.data, link.len);
    self->setLink(link_QString);
}

void Attica__DownloadDescription_SetDistributionType(Attica__DownloadDescription* self, const libqt_string distributionType) {
    QString distributionType_QString = QString::fromUtf8(distributionType.data, distributionType.len);
    self->setDistributionType(distributionType_QString);
}

void Attica__DownloadDescription_SetPriceReason(Attica__DownloadDescription* self, const libqt_string priceReason) {
    QString priceReason_QString = QString::fromUtf8(priceReason.data, priceReason.len);
    self->setPriceReason(priceReason_QString);
}

void Attica__DownloadDescription_SetPriceAmount(Attica__DownloadDescription* self, const libqt_string priceAmount) {
    QString priceAmount_QString = QString::fromUtf8(priceAmount.data, priceAmount.len);
    self->setPriceAmount(priceAmount_QString);
}

void Attica__DownloadDescription_SetSize(Attica__DownloadDescription* self, unsigned int size) {
    self->setSize(static_cast<uint>(size));
}

void Attica__DownloadDescription_SetGpgFingerprint(Attica__DownloadDescription* self, const libqt_string fingerprint) {
    QString fingerprint_QString = QString::fromUtf8(fingerprint.data, fingerprint.len);
    self->setGpgFingerprint(fingerprint_QString);
}

void Attica__DownloadDescription_SetGpgSignature(Attica__DownloadDescription* self, const libqt_string signature) {
    QString signature_QString = QString::fromUtf8(signature.data, signature.len);
    self->setGpgSignature(signature_QString);
}

void Attica__DownloadDescription_SetPackageName(Attica__DownloadDescription* self, const libqt_string packageName) {
    QString packageName_QString = QString::fromUtf8(packageName.data, packageName.len);
    self->setPackageName(packageName_QString);
}

void Attica__DownloadDescription_SetRepository(Attica__DownloadDescription* self, const libqt_string repository) {
    QString repository_QString = QString::fromUtf8(repository.data, repository.len);
    self->setRepository(repository_QString);
}

void Attica__DownloadDescription_SetTags(Attica__DownloadDescription* self, const libqt_list /* of libqt_string */ tags) {
    QList<QString> tags_QList;
    tags_QList.reserve(tags.len);
    libqt_string* tags_arr = static_cast<libqt_string*>(tags.data);
    for (size_t i = 0; i < tags.len; ++i) {
        QString tags_arr_i_QString = QString::fromUtf8(tags_arr[i].data, tags_arr[i].len);
        tags_QList.push_back(tags_arr_i_QString);
    }
    self->setTags(tags_QList);
}

libqt_string Attica__DownloadDescription_Version(const Attica__DownloadDescription* self) {
    QString _ret = self->version();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void Attica__DownloadDescription_SetVersion(Attica__DownloadDescription* self, const libqt_string version) {
    QString version_QString = QString::fromUtf8(version.data, version.len);
    self->setVersion(version_QString);
}

void Attica__DownloadDescription_Delete(Attica__DownloadDescription* self) {
    delete self;
}
